### DISCUSSION TITLE
My experience with riak_core

### DISCUSSION THREAD

**Discussion by jessestimpson** [OpenRiak/.github#7] [Show and tell]
Hello, at some point last year, being unaware of the OpenRiak effort, I became interested in seeing whether or not riak_core could be used to build a general purpose crdt library for Elixir. I did make some progress, although not nearly enough to make any releases.

After seeing the HN post today, I figured if anyone out there may be interested, they would be here. 😄 I haven't touched this work in over a year, so there's a lot I've forgotten, but I'll do my best to remember what I did.

My [experimental riak_core branch](https://github.com/JesseStimpson/riak_core/tree/experimental) has [some aggressive changes](https://github.com/JesseStimpson/riak_core/blob/experimental/CHANGES.md) in it. IIRC I spent the bulk of my time removing the leveldb dependency. I was pretty ruthless about ripping out code that was inconvenient at the time, so you'll find a lot missing, and I'm sure I broke some important things. 

However, I was successful in getting a [multinode integration test](https://github.com/JesseStimpson/riak_core/blob/experimental/test/key_value_SUITE.erl) with a [basic in-memory key-value vnode](https://github.com/JesseStimpson/riak_core/blob/experimental/test/key_value_example.erl).

Then, I began playing around with an [Elixir API on the riak_core CRDTs](https://github.com/JesseStimpson/crdt). My interest in the project fizzled, and I haven't looked at it since.

Here's a trivial example that demonstrates the Elixir API. I admit it's way too pipe heavy, but I hope it gets the point across.

```elixir
:ok =
  Crdt.for(:incr_api)
  |> Crdt.at(:latest)
  |> Crdt.increment(:like)
  |> Crdt.apply()

{:ok, %{latest: %{like: 1}}} = Crdt.get(:incr_api)

:ok =
  Crdt.for(:incr_api)
  |> Crdt.at(:latest)
  |> Crdt.decrement(:like)
  |> Crdt.apply()

{:ok, %{latest: %{like: 0}}} = Crdt.get(:incr_api)
```

Anyway, I just wanted to introduce myself since I've been on a team that has had success in using riak at scale in production, and I believe there lives some excellent engineering work inside the riak repos that is begging for a practical modern application.

Please let me know if I can be any help to the project. In particular, is there any small work in riak_core that I can get my feet wet on?

Jesse

**Comment by martinsumner** [comment]
Hello!

I don't know if you came across [riak-core-lite](https://riak-core-lite.github.io/) which had similar goals.

There is some work which needs doing on riak_core, but at the moment it isn't high up the priority list.

- There are [a couple of PRs which are stalled](https://github.com/OpenRiak/riak_core/pulls), partly due to relative importance, but also as we're not sure this is the right approach to achieving their goals.
- There's a [general need to strip under-utilised things out](https://github.com/OpenRiak/riak_core/issues/5), as you've discovered.
- At some stage we want to remove the bespoke connection setup for handoff (it is a legacy of network issues in older OTP versions).
- Also In the long-term project list is further optimisation of coverage planning, to [extend the previous work](https://github.com/basho/riak_kv/issues/1801) to get sub 1ms planning a larger ring sizes.

Away from riak_core, @adkron was thinking of looking at the Elixir client for Riak, which hasn't had much attention in the past few years, it may be worth connecting with Amos.

CRDTs and riak_dt is another thing that we've struggled to give sufficient attention to in recent years.

Help is welcome.  If you're an erlef member there's an open-riak channel in the EEF slack.  I'll let everyone there know you're here.

**Reply by adkron** [reply]
I really need to prioritize some time to set up a stack and do this. I'd also like to work on a gleam driver.
